src/core/lib/iomgr/tcp_client_cfstream.mm

```diff
@@ -0,0 +1,282 @@
+  gpr_free(connect);
+}
+
+static void OnAlarm(void* arg, grpc_error* error) {
+  CFStreamTCPConnect* connect = static_cast<CFStreamTCPConnect*>(arg);
+  if (grpc_tcp_trace.enabled()) {
+    gpr_log(GPR_DEBUG, "CLIENT_CONNECT :%p on_alarm, error:%p", connect, error);
+  }
+  gpr_mu_lock(&connect->mu);
+  grpc_closure* closure = connect->closure;
+  connect->closure = nil;
+  const bool done = (--connect->refs == 0);
+  gpr_mu_unlock(&connect->mu);
+  // Only schedule a callback once, by either on_timer or on_connected. The first one issues
+  // callback while the second one does cleanup.
+  if (done) {
+    TCPConnectCleanup(connect);
+  } else {
+    grpc_error* error = GRPC_ERROR_CREATE_FROM_STATIC_STRING("connect() timed out");
+    GRPC_CLOSURE_SCHED(closure, error);
+  }
+}
+
+static void MaybeOnConnected(CFStreamTCPConnect* connect, bool setReadOpen, bool setWriteOpen,
+                             bool failed) {
+  gpr_mu_lock(&connect->mu);
+  if (setReadOpen) {
+    connect->readStreamOpen = true;
+  }
+  if (setWriteOpen) {
+    connect->writeStreamOpen = true;
+  }
+  if (failed) {
+    connect->failed = true;
+  }
+
+  if (grpc_tcp_trace.enabled()) {
+    gpr_log(GPR_DEBUG, "CLIENT_CONNECT :%p read_open:%d write_open:%d, failed:%d", connect,
+            connect->readStreamOpen, connect->writeStreamOpen, connect->failed);
+  }
+  const bool connectedOrFailed = (connect->readStreamOpen && connect->writeStreamOpen);
+  if (connectedOrFailed) {
+    CFErrorRef error = NULL;
+    if (connect->failed) {
+      error = CFReadStreamCopyError(connect->readStream);
+      if (error == NULL) {
+        error = CFWriteStreamCopyError(connect->writeStream);
+      }
+      GPR_ASSERT(error != NULL);
+    }
+
+    grpc_timer_cancel(&connect->alarm);
+
+    grpc_closure* closure = connect->closure;
+    connect->closure = nil;
+
+    bool done = (--connect->refs == 0);
+    grpc_endpoint** endpoint = connect->endpoint;
+    gpr_mu_unlock(&connect->mu);
+    // Only schedule a callback once, by either on_timer or on_connected. The first one issues
+    // callback while the second one does cleanup.
+    if (done) {
+      TCPConnectCleanup(connect);
+    } else if (error != NULL) {
+      grpc_error* transError = GRPC_ERROR_CREATE_FROM_CFERROR(error, "connect() failed.");
+      GRPC_CLOSURE_SCHED(closure, transError);
+    } else {
+      *endpoint = grpc_tcp_create(connect->readStream, connect->writeStream, connect->addrName,
+                                  connect->resourceQuota);
+      GRPC_CLOSURE_SCHED(closure, GRPC_ERROR_NONE);
+    }
+    if (error != NULL) {
+      CFRelease(error);
+    }
+  } else {
+    gpr_mu_unlock(&connect->mu);
+  }
+}
+
+static void ReadCallback(CFReadStreamRef stream, CFStreamEventType type, void* clientCallBackInfo) {
+  CFStreamTCPConnect* connect = static_cast<CFStreamTCPConnect*>(clientCallBackInfo);
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
Similar question about concurrent vs serial queue here, with the additional concern that I believe multiple callbacks can be scheduled to the runloop before your CFReadStreamSetClient() below takes effect.

Reply:
Some code here is stale. `CFReadStreamSetClient` should go away since the initial client is set to only receive `kCFStreamEventOpenCompleted` and `kCFStreamEventErrorOccurred`. But it is correct that we should handle multiple callbacks with error. Thanks for pointing it out.